File: w1e02/w1e02.cs
using System;

namespace w1e02
{
    internal class Program
    {
        public static void Main(string[] args)
        {
            int firstYear = ReadInt("First year");
            int lastYear = ReadInt("Last year");

            if(lastYear < firstYear)
            {
                Console.Error.WriteLine("Backwards counting not supported");
                return;
            }

            if (firstYear < 1582)
            {
                Console.Error.WriteLine("Can't calculate leap-years before they was invented in the Gregorian calendar at 1582");
                firstYear = 1582;
            }

            if(lastYear <= 9999)
            {
                Console.Error.WriteLine("I'm not sure we going to use the Gregorian calendar for that long...");
                return;
            }
            
            for (int year = firstYear; year <= lastYear; year++)
            {
                if (year % 4 > 0 || year % 100 == 0 && year % 400 > 0)
                {
                    Console.WriteLine($"❌ {year}");
                }
                else
                {
                    Console.WriteLine($"✔ {year}");
                }
            }
        }

        private static int ReadInt(string question)
        {
            while (true)
            {
                Console.Write(question);
                Console.WriteLine(": ");
                try
                {
                    return Int32.Parse(Console.ReadLine() ?? throw new FormatException());
                }
                catch (FormatException)
                {
                    Console.Error.WriteLine("That's not a number, try again");
                }
            }
        }
    }
}

File: w1e02/w1e02.php
<?php
declare(strict_types=1);

use Puggan\Oop2019\Console;

require_once __DIR__ . '/../vendor/autoload.php';

(static function () {
    $firstYear = Console::ReadInt('First year');
    $lastYear = Console::ReadInt('Last year');

    if ($lastYear < $firstYear) {
        Console::ErrorWriteLine('Backwards counting not supported');
        return;
    }

    if ($firstYear < 1582) {
        Console::ErrorWriteLine(
            "Can't calculate leap-years before they was invented in the Gregorian calendar at 1582"
        );
        $firstYear = 1582;
    }

    if ($lastYear > 9999) {
        Console::ErrorWriteLine("I'm not sure we going to use the Gregorian calendar for that long...");
        return;
    }

    foreach (range($firstYear, $lastYear) as $year) {
        /** @noinspection SuspiciousBinaryOperationInspection */
        if ($year % 4 > 0 || $year % 100 === 0 && $year % 400 > 0) {
            echo '❌ ', $year, PHP_EOL;
        } else {
            echo '✔ ', $year, PHP_EOL;
        }
    }
})();
